Improve the github bridge

samoht opened the pull request.

samoht commented on src/vgithub/vgithub.ml:
```diff
@@ -781,30 +788,36 @@ module Sync (API: API) (DK: Datakit_S.CLIENT) = struct
-      ok ()
+          of_commit c >>*= fun priv ->
+          of_tree (E ((module DK.Transaction), tr)) >>*= fun pub ->
+          (if compare_snapshot old_t.priv priv = 0 then
```
@talex5 I think this does the right thing: e.g. it merges only if the private branch has changed.

talex5 replied:
Looks right. I'll add an issue to support this directly in DataKit though.

talex5 commented on hooks/datakit-gh-hooks/server/event.go:
```diff
@@ -112,7 +117,7 @@ func (h *Server) CommitDir(e github.StatusEvent) ([]string, error) {
 
 	h.logger.Debugf("user=%s, repo=%s", user, repo)
-	commitDir := []string{user, repo, "commits", sha}
+	commitDir := []string{user, repo, "commit", "status", sha}
```
Should this be named `statusDir` then?

talex5 commented on tests/test_github.ml:
```diff
@@ -240,6 +239,11 @@ let test_events dk =
+  VG.sync ~policy:`Once s ~priv ~pub ~token:t >>= fun s ->
+  VG.sync ~policy:`Once s ~priv ~pub ~token:t >>= fun s ->
+  VG.sync ~policy:`Once s ~priv ~pub ~token:t >>= fun s ->
+  VG.sync ~policy:`Once s ~priv ~pub ~token:t >>= fun s ->
```
Is this intentional or a copy-and-paste error?

samoht replied:
it's intentional, I want to really be sure that no actions is taken :-)

samoht commented:
I've renamed the variable to `statusDir` and rebased to master. I have also added 2 more commits:
- avoid modifying the whole world on startup for it to look like `github-metadata`. It's just confusing, and even more if it's revert back some changes made by the users. It also seems to change the owner of every "status" on the open PRs, which can be confusing as well.
- add a `--dry-updates` command-line argument to the GH bridge, to run it in "read-only" mode.

samoht commented:
There is still one missing commit that I would like to make:

- would be great to create only one transaction in the private branch too instead of make each phases create in a separate commit, while at the same time the GH webhook continue to write in the branch (so possible races there too).

Anyway, this is not a blocker, so I'll merge the PR, try to deploy it and make another PR with that fix later.

talex5 commented on src/vgithub/vgithub.mli:
```diff
@@ -118,12 +118,14 @@ module Sync (API: API) (DK: Datakit_S.CLIENT): sig
     t -> t Lwt.t
 (** [sync t ~pub ~priv ~token] mirror GitHub changes in the DataKit
     public branch [pub]. It uses the private branch [priv] to store
     the received wehook event states. It connects to the GitHub API
```
Typo: `wehook`

talex5 commented on src/vgithub/vgithub.ml:
```diff
@@ -549,82 +547,121 @@ module Sync (API: API) (DK: Datakit_S.CLIENT) = struct
-  let empty =
-    { hooks = HookSet.empty; user_repos = UserRepoSet.empty;
-      tree = None; commit = None }
+  type t = { pub: snapshot; priv: snapshot; synced: bool }
```
Would be good to have a comment here saying what `synced` means.

talex5 commented on src/vgithub/vgithub.ml:
```diff
@@ -770,41 +785,49 @@ module Sync (API: API) (DK: Datakit_S.CLIENT) = struct
-      ok ()
-    in
+    let call_github_api old_t current =
+      if not t.synced then ok ()
```
Assuming `synced` means GitHub is already up-to-date, this test seems to be the wrong way around.

samoht commented:
@talex5 I've renamed `synced` into `merged` and try to add some comments. Is that clearer?

talex5 commented:
Yes, thanks!
